docs/devel/api_changes.md

```diff
@@ -0,0 +1,237 @@
+
+## Changing the internal API
+
+Now it is time to change the internal API so your versioned changes can be
+used.
+
+### Edit types.go
+
+Similar to the versioned APIs, the struct definitions for the internal API are
+in `pkg/api/types.go`.  Edit those files to reflect the change you want to
+make.  Keep in mind that the internal API must be able to express *all* of the
+versioned APIs.
+
+## Edit validation.go
+
+Most changes made to the APIs need some form of input validation.  Validation
+is currently done on internal objects in `pkg/api/validation/validation.go`.
+This validation is the one of the first opportunities we have to make a great
+user experience - good error messages and thorough validation help ensure that
+users are giving you what you expect and, when they don't, that they know why
+and how to fix it.  Think hard about the contents of `string` fields, the bounds
+of `int` fields and the requiredness/optionalness of fields.
+
+Of course, code needs tests - `pkg/api/validation/validation_test.go`.
+
+## Edit version conversions
+
+At this point you have both the versioned API changes and the internal API
+changes done.  If there are any notable differences - field names, types,
+structural change in particular - you must add some logic to convert versioned
+APIs to and from internal.  This conversion process is a "star" with the
+internal API at the center - every versioned API must be able to convert to the
+internal API, and vice-versa, but versioned APIs do not convert to other
+versioned APIs directly.  This sounds like a heavy process, but in reality we
+don't intend to keep more than a small number of versions alive at once.
+
+If you see errors from the `serialization_test`, it may indicate the need for
+explicit conversions.
+
+The conversion code resides with each versioned API -
+`pkg/api/<version>/conversion.go.  Unsurprisingly, this also requires you to
+add tests to `pkg/api/<version>/conversion_test.go`.
+
+## Update the fuzzer
+
+Part of our testing regimen for APIs is to "fuzz" (fill with random values) API
+objects and then convert them to and from the different API versions.  This is
+a great way of exposing places where you lost information or made bad
+assumptions.  If you have added any fields which need very careful formatting
+(the test does not run validation) or if you have made assumptions such as
+"this slice will always have at least 1 element", you may get an error or even
+a panic from the `serialization_test`.  If so, look at the diff it produces (or
+the backtrace in case of a panic) and figure out what you forgot.  Encode that
+into the fuzzer's custom fuzz functions.
+
+The fuzzer can be found in `pkg/api/testing/fuzzer.go`.
+
+## Update the semantic comparisons
+
+VERY VERY rarely is this needed, but when it hits, it hurts.  In some rare
+cases we end up with objects (e.g. resource quantites) that have morally
+equivalent values with different bitwise representations (e.g. value 0 with a
```

Comment:
0 isn't a good example. How about 10?

Note that we recently ran into a problem with this approach: readonly and non-updatable fields. If a client does an update that would change the representation but not the semantic value of a readonly or non-updatable field, the update currently fails.

An alternative would be normalization. That's the solution we plan to use for creationTimestamp, because the system sets that field in the first place. We currently have a policy of not altering the representation or value of fields set by the client/user.